new-log-viewer/src/services/LogFileManager.ts

```diff
@@ -0,0 +1,200 @@
+     *
+     * @param cursor The cursor indicating the page to load. See {@link CursorType}.
+     * @return An object containing the logs as a string, a map of line numbers to log event
+     * numbers, and the line number of the first line in the cursor identified event.
+     * @throws {Error} if the `#loadFile` method has not been called before this method.
+     */
+    loadPage (cursor: CursorType): {
+        logs: string,
+        beginLineNumToLogEventNum: BeginLineNumToLogEventNumMap,
+        cursorLineNum: number
+    } {
+        console.debug(`loadPage: cursor=${JSON.stringify(cursor)}`);
+        if (null === this.#decoder) {
+            throw new Error("#loadFile() must be first called.");
+        }
+
+        const {beginLogEventNum, endLogEventNum} = this.#getCursorRange(cursor);
+        const results = this.#decoder.decode(beginLogEventNum - 1, endLogEventNum);
+        if (null === results) {
+            throw new Error("Error occurred during decoding. " +
+                `beginLogEventNum=${beginLogEventNum}, ` +
+                `endLogEventNum=${endLogEventNum}`);
+        }
+
+        const messages: string[] = [];
+        const beginLineNumToLogEventNum: BeginLineNumToLogEventNumMap = new Map();
+        let currentLine = 1;
+        results.forEach((r) => {
+            const [
+                msg,
+                ,
+                ,
+                logEventNum,
+            ] = r;
+
+            messages.push(msg);
+            beginLineNumToLogEventNum.set(currentLine, logEventNum);
+            currentLine += msg.split("\n").length - 1;
+        });
+
+        return {
+            logs: messages.join(""),
+            beginLineNumToLogEventNum: beginLineNumToLogEventNum,
+            cursorLineNum: 1,
+        };
+    }
+
+    /**
+     * Loads a file from a given source and decodes it using the appropriate decoder based on the
+     * file extension.
+     *
+     * @param fileSrc The source of the file to load. This can be a string representing a URL, or a
+     * File object.
+     * @return A promise that resolves with the number of log events found in the file.
+     * @throws {Error} If the file source type is not supported.
+     */
+    async #loadFile (fileSrc: FileSrcType) {
+        if ("string" === typeof fileSrc) {
+            this.#fileName = getBasenameFromUrlOrDefault(fileSrc);
+            this.#fileData = await getUint8ArrayFrom(fileSrc, () => null);
+        } else {
+            // TODO: support file loading via Open / Drag-n-drop
+            throw new Error("Read from file not yet supported");
+        }
+    }
+
+    /**
+     * Constructs a decoder instance based on the file extension.
+     *
+     * @param decoderOptions Initial decoder options.
+     * @throws {Error} if #fileName or #fileData hasn't been init, or a decoder cannot be found.
+     */
+    #initDecoder = (decoderOptions: DecoderOptionsType): void => {
```

Comment:
Fwiw, this method could be taken out of the class since it's really only meant to be called once during construction. That would also allow us to get of the `Unexpected usage` exception. That said, it's not a huge deal to leave it in the class.

Reply:
Instead, the method is kept to retain the side effects, but now it returns the constructed decoder so that the `#decoder` instance can be definitely assigned in LogFileManager's constructor.